src/cc/libbpf.c

```diff
@@ -70,13 +72,63 @@
 #define PERF_FLAG_FD_CLOEXEC (1UL << 3)
 #endif
 
+#define BCC_INSTANCE_FAIL -1
+#define BCC_INSTANCE_NOT_INIT -2
+#define BCC_INSTANCE_USE_DEFAULT -3
+static int bcc_instance_probe_cnt = BCC_INSTANCE_NOT_INIT;
+
 static int probe_perf_reader_page_cnt = 8;
 
 static uint64_t ptr_to_u64(void *ptr)
 {
   return (uint64_t) (unsigned long) ptr;
 }
 
+static int bcc_get_instance(char *buf, size_t buf_size) {
+  int res;
+  // An instance has already been created.
+  if (bcc_instance_probe_cnt >= 0) {
+    res = snprintf(buf, buf_size, "/sys/kernel/debug/tracing/instances/bcc_%d/", getpid());
+    if (res <= 0 || res >= buf_size)
+      return BCC_INSTANCE_FAIL;
+    return 0;
+  }
+  // Already failed creating an instance. Don't try again and use default.
+  if (bcc_instance_probe_cnt == BCC_INSTANCE_USE_DEFAULT)
+    goto default_instance;
+  // Try create an instance.
+  if (access("/sys/kernel/debug/tracing/instances/", F_OK) != 0) {
+    // Fail if instance folder does not exist.
+    bcc_instance_probe_cnt = BCC_INSTANCE_USE_DEFAULT;
+    goto default_instance;
+  }
+  res = snprintf(buf, buf_size, "/sys/kernel/debug/tracing/instances/bcc_%d/", getpid());
+  if (res <= 0 || res >= buf_size)
```

Comment:
We have numerous cases like 
  res = snprintf(..)
  if (res <= 0 || res >= buf_size) {
    ...
  }

Can we abstract this out as a macro or static function? This way, code is more concise and less distraction?